src/test/java/io/kamax/matrix/client/regular/MatrixHttpClientTest.java

```diff
@@ -39,6 +50,95 @@ public MatrixHttpClientTest() throws URISyntaxException {
     }
 
     @Test
+    public void loginWithDeviceId() throws URISyntaxException {
+        stubFor(post(urlEqualTo(loginUrl))
+                .withRequestBody(equalToJson("{\"type\": \"m.login.password\"," + //
+                        "\"user\": \"" + user.getLocalPart() + "\"," + //
+                        "\"password\": \"" + password + "\"," + //
+                        "\"device_id\": \"" + deviceId + "\"}"))
+                .willReturn(aResponse().withStatus(200)
+                        .withBody("{\"user_id\": \"" + user.getId() + "\"," + //
+                                "\"access_token\": \"" + testToken + "\"," + //
+                                "\"home_server\": \"" + new MatrixHomeserver(domain, baseUrl) + "\"," + //
+                                "\"device_id\": \"" + deviceId + "\"}")));
+
+        MatrixHomeserver hs = new MatrixHomeserver(domain, baseUrl);
+        MatrixClientContext context = new MatrixClientContext(hs, deviceId);
+        MatrixHttpClient client = new MatrixHttpClient(context);
+
+        MatrixPasswordLoginCredentials credentials = new MatrixPasswordLoginCredentials(user.getLocalPart(), password);
+        client.login(credentials);
+
+        assertEquals(deviceId, client.getDeviceId().get());
+    }
```

Comment:
What about validating the required fields are not blank (`StringUtils.isNotBlank()`) as per spec:
- Matrix ID
- Access token
- Device ID
- Homeserver domain